### PR TITLE
[GSoC] Multi-factor feature for RubyGems.

### PR DESCRIPTION
**Pull request by ecnelises**
# Description:

Hello. This is my GSoC project, dedicated to add multifactor authentication to both RubyGems command program and Gemcutter ([RubyGems.org](https://rubygems.org)).

Work for the Gemcutter part has almost been finished (see [PR #1753](https://github.com/rubygems/rubygems.org/pull/1753), [PR #1729](https://github.com/rubygems/rubygems.org/pull/1729) and a series of [my progress reports](https://ecnelises.github.io/)).

## Content:

This PR will contain my changes to RubyGems client, adding multifactor auth for `gem push`, `gem signin` and `gem owner` commands. Since no command for editing profile, adding command for changing multifactor auth settings seems unnecessary.

## Workflow:

- User set up multifactor auth well in the site. (into `mfa_login_and_write` level)
- When user does the actions requiring MFA, an OTP prompt is shown. Or user can add `--otp` option into command, like `gem push mygem-0.0.0.gem --otp 123456`.
- If the OTP is incorrect, operation fails with failure text.
______________

# Tasks:

- [x] Add OTP requirement to `push_command`.
- [x] Add OTP requirement to `owner_command`.
- [x] Add OTP prompt to `sign_in`.
- [ ] Support for `yank_command`.
- [x] Write related tests.

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

### CODE REVIEW

File: lib/rubygems/gemcutter_utilities.rb [hunk 1]
```diff
@@ -119,6 +119,40 @@ def sign_in sign_in_host = nil
...
+  # OTP code is needed.
+
+  def need_otp?
+    return false if options[:suppress_mfa]
+    unless instance_variable_defined? :@mfa_level
+      response = rubygems_api_request(:get, 'api/v1/multifactor_auth') do |request|
+        request.add_field 'Authorization', api_key
+      end
+
+      # For compatibility to Gemcutters without mfa support
```
Comment (hsbt) [inline lib/rubygems/gemcutter_utilities.rb]: `Gemcutters` is obsoleted name. Please use `rubygems.org`

**Comment by colby-swandale** [comment]
I don't like the idea of the use of `otp` and `mfa`, the code would be clearer if you stick with just the one name.

**Comment by colby-swandale** [comment]
Using both `otp` and `mfa` does not seem clean, the code would be clearer if you stick with just the one name.

File: lib/rubygems/gemcutter_utilities.rb [hunk 2]
```diff
@@ -119,6 +119,40 @@ def sign_in sign_in_host = nil
...
+                   end
+    end
+
+    @mfa_level == 'mfa_login_and_write'
+  end
+
+  ##
+  # Require user for extra OTP code if multifactor authentication is enabled.
+
+  def run_mfa_check
```
Comment (colby-swandale) [inline lib/rubygems/gemcutter_utilities.rb]: `run` doesn't really mean anything. Maybe just `mfa_check`?

**Comment by colby-swandale** [comment]
Can you please explain how the user flows works with 2FA. Looking at the code, does the user need to sign in with 2FA first or an error will be given? 

Can we just ask the user to give their MFA token whenever they sign in to an account which has 2FA?

**Comment by ecnelises** [comment]
@colby-swandale 
Thanks. _mfa_ is better known than _otp_, mixing them would confuse people not familiar with how mfa/2fa works.

Here we have a concept about _level of mfa_, inspired by npm. A user can set his level to `login_only` or `login_and_write`. `login_only` means when a user does login, a page requiring extra digit code will be shown. `login_and_write` also requires the digit code for some important operations besides login. Since I have implemented digit check for website login, add a check for MFA code to client seems more reasonable than not.

**Comment by ecnelises** [comment]
A lot of `gem` commands need to be sign in. But if we require user input code first, operations next may also require digit code. That will make operation fail. Or retieving API key does not need multifactor auth as long as we require digit code for all important operations.

File: lib/rubygems/gemcutter_utilities.rb [hunk 3]
```diff
@@ -119,6 +119,40 @@ def sign_in sign_in_host = nil
...
+      # For compatibility to Gemcutters without mfa support
+      @mfa_level = case response
+                   when Net::HTTPNotFound
+                     'no_mfa'
+                   else
+                     with_response(response) { |resp| resp.body }
+                   end
+    end
+
+    @mfa_level == 'mfa_login_and_write'
```
Comment (tjouan) [inline lib/rubygems/gemcutter_utilities.rb]: If there isn't a many different values for `@mfa_level`, shouldn't symbols be
favored over strings here?

File: lib/rubygems/gemcutter_utilities.rb [hunk 4]
```diff
@@ -130,7 +130,7 @@ def need_otp?
         request.add_field 'Authorization', api_key
       end
 
-      # For compatibility to Gemcutters without mfa support
+      # For compatibility to servers without mfa support
```
Comment (tjouan) [inline lib/rubygems/gemcutter_utilities.rb]: I think this change belongs to another commit (the one adding this comment).

File: lib/rubygems/commands/push_command.rb [hunk 5]
```diff
@@ -38,13 +38,18 @@ def initialize
       options[:host] = value
     end
 
+    add_option('--mfa CODE', 'Digit code for multifactor authentication') do |value, options|
```
Comment (segiddins) [inline lib/rubygems/commands/push_command.rb]: can this repeated code be factored into a module?

File: lib/rubygems/gemcutter_utilities.rb [hunk 6]
```diff
@@ -119,6 +119,40 @@ def sign_in sign_in_host = nil
...
   end
 
+  ##
+  # Fetch user's multifactor authentication settings and return if extra digit
+  # code is needed.
+
+  def need_mfa?
+    return false if options[:suppress_mfa]
+    unless instance_variable_defined? :@mfa_level
+      response = rubygems_api_request(:get, 'api/v1/multifactor_auth') do |request|
```
Comment (sonalkr132) [inline lib/rubygems/gemcutter_utilities.rb]: Won't this be adding additional api call to all requests? For default case, we should assume that mfa on api is disabled. mfa is an opt-in feature.
From server, we can return 403 or something more appropriate for cases where otp was required but not provided. Only then, client will prompt for otp and repeat the request.

Comment (ecnelises) [reply lib/rubygems/gemcutter_utilities.rb]: Here `options[:supress_mfa]` is just an internal option for keep compatibility with older test cases, since many of them relies on `last_request`. And fail message the client received is `plain/text`, should I just match it literally?

File: lib/rubygems/commands/push_command.rb (hunk 5 above)
Comment (ecnelises) [reply lib/rubygems/commands/push_command.rb]: Yes. Thanks.

File: lib/rubygems/gemcutter_utilities.rb (hunk 6 above)
Comment (sonalkr132) [reply lib/rubygems/gemcutter_utilities.rb]: > for keep compatibility with older test cases, since many of them relies on last_request

Can you please explain more about this limitation? Are you sure we can't do away with `options[:suppress_mfa]`?

>  should I just match it literally?

Yeah, checking for response text would be good enough unless we come up with something more elegant.

Comment (ecnelises) [reply lib/rubygems/gemcutter_utilities.rb]: If we want to remove `options[:suppress_mfa]`, I have tried that create a new module dynamically, setting `need_mfa?` always return false, and prepend it to `Gem::Commands::PushCommand` in test cases which will complain if mfa related requests are added. This method should be not so intrusive, like [here](https://paste.ofcode.org/3aBhsqT54rrcmeZUfjsBdTh).

File: lib/rubygems/gemcutter_utilities.rb (hunk 3 above)
Comment (tjouan) [reply lib/rubygems/gemcutter_utilities.rb]: No idea, I didn't even thought about performance when I commented to be honest.
It was more about "semantics", that it seemed used as an "identifier" more than
a string.
But on the other hand, as you said this data is received as string already, if
the spec pass in theory we shouldn't need more code :-)
Anyway it was a small detail I just asked about in case it would help, feel
free to ignore.

File: lib/rubygems/commands/owner_command.rb [hunk 7]
```diff
@@ -87,6 +88,16 @@ def manage_owners method, name, owners
...
+          request.add_field "OTP", options[:mfa] if options[:mfa]
+        end
+
+        if need_mfa? response
+          check_mfa
+          response = rubygems_api_request method, "api/v1/gems/#{name}/owners" do |request|
+            request.set_form_data 'email' => owner
+            request.add_field "Authorization", api_key
+            request.add_field "OTP", options[:mfa]
+          end
```
Comment (sonalkr132) [inline lib/rubygems/commands/owner_command.rb]: Please don't duplicate code. Make a method out of common code where otp is optional argument.
Also, I agree with @colby-swandale 
> the code would be clearer if you stick with just the one name.

`"OTP", options[:mfa]` :no_entry_sign: 
Please use `otp` for everything.

File: lib/rubygems/gemcutter_utilities.rb [hunk 8]
```diff
@@ -113,12 +123,31 @@ def sign_in sign_in_host = nil
       request.basic_auth email, password
     end
 
+    if need_mfa? response
+      check_mfa
```
Comment (sonalkr132) [inline lib/rubygems/gemcutter_utilities.rb]: `check_mfa` and `need_mfa?` always seem to go together. Can we please call `check_mfa` from `need_mfa?` instead?

File: lib/rubygems/gemcutter_utilities.rb [hunk 9]
```diff
@@ -156,6 +174,20 @@ def with_response response, error_prefix = nil
...
   end
 
+  ##
+  # Returns whether the user has enabled multifactor authentication from
+  # +response+ text.
+
+  def need_otp? response
+    return unless response.kind_of?(Net::HTTPUnauthorized) &&
+        response.body.start_with?('You have enabled multifactor authentication')
+    unless options[:otp]
```
Comment (sonalkr132) [inline lib/rubygems/gemcutter_utilities.rb]: `return true if options[:otp]`

**Comment by ecnelises** [comment]
😄 Is there any problems for current implementation status? I wrote in detail about what it will look like and how it's implemented in [GSoC final report](https://gist.github.com/ecnelises/9654e59877aa336977c1409ef540e2a9). Sorry for out of work these days, maybe I should add 2FA to `yank` now.

**Comment by hsbt** [comment]
@indirect @sonalkr132 Can you handle this issue with rubygems.org side. I'm okay to merge this pull-request targeted RubyGems 3.0.

**Comment by hsbt** [comment]
@eileencodes Do you still waiting to review this?

**Comment by ChungTranThanh** [comment]
https://github.com/eileencodes

**Comment by sonalkr132** [comment]
@hsbt sorry about delay with this. We did have PR on rubygems.org side blocking it which is merged now. 
Functionality wise, this PR is complete. I only need to test these changes with rubygems.org prod api, to ensure we don't break any existing functionality (which I hope to get done by EOW). You are also welcome to test it out. you can enabled mfa related featured on https://rubygems.org by running `document.cookie = 'mfa_feature=true;path=/'` in your browser console.

**Comment by hsbt** [comment]
@sonalkr132 Thank you for your explanation!

RG 3.0 is a good time to merge this feature. The ruby core team will release Ruby 2.6.0-rc1 with RG 3.0 that was the feature freeze version at early December 2018.

**Comment by sonalkr132** [comment]
> need to test these changes with rubygems.org prod api, to ensure we don't break any existing functionality

verified that gem `owner`, `signin` and `push` work as intended with `ui only` and `ui and api` mfa levels. Feel free to way in on aesthetic of PR, specially part we will show to users. 
@ecnelises has writes guides for [setting up mfa](https://github.com/rubygems/guides/blob/ccbd9c778490d8f596fbeda9e433daaa33b1f5df/setting-up-multifactor-authentication.md) and [cli usage](https://github.com/rubygems/guides/blob/ccbd9c778490d8f596fbeda9e433daaa33b1f5df/using-mfa-in-command-line.md), where you can find further details (PR on https://github.com/rubygems/guides/pull/218).

**Comment by hsbt** [comment]
@bundlerbot r+

**Comment by hsbt** [comment]
@ecnelises @sonalkr132 Thanks for your work. I'm going to release this feature with RG 3.0 at 25, Dec 2018.

**Comment by sonalkr132** [comment]
> - [ ] Support for yank_command.

Not really sure why @ecnelises decided to keep it separate but PR for it is now merged on rubygems.org, I guess now he can add it here.

**Comment by ecnelises** [comment]
@sonalkr132 I thought that may delay the review process and I’ve waited for `yank` related API on rubygems.org for a while. I can add it in future PRs. :-) Thanks.

邱超凡

________________________________
寄件者: Aditya Prakash <notifications@github.com>
寄件日期: 星期六, 12月 1, 2018 11:08 下午
收件者: rubygems/rubygems
副本: Qiu Chaofan; Mention
主旨: Re: [rubygems/rubygems] [GSoC] Multi-factor feature for RubyGems. (#2369)


  *    Support for yank_command.

Not really sure why @ecnelises<https://github.com/ecnelises> decided to keep it separate but PR for it is now merged on rubygems.org, I guess now he can add it here.

**Comment by sonalkr132** [comment]
yeah, its cool. yank is not as often used (although very important for mfa) command.
If @hsbt really wants to see make it in one go, I hope he will be kind enough to merge it again :pray:
